libsolidity/parsing/Parser.cpp

```diff
@@ -242,14 +245,31 @@ ASTPointer<ImportDirective> Parser::parseImportDirective()
 	return nodeFactory.createNode<ImportDirective>(path, unitAlias, move(symbolAliases));
 }
 
-ContractDefinition::ContractKind Parser::parseContractKind()
+std::pair<ContractDefinition::ContractKind, bool> Parser::parseContractKind()
 {
 	ContractDefinition::ContractKind kind;
+	bool doNext = true;
+	bool abstract = false;
 	switch(m_scanner->currentToken())
 	{
 	case Token::Interface:
 		kind = ContractDefinition::ContractKind::Interface;
 		break;
+	case Token::Abstract:
+		if (Token::Abstract == m_scanner->peekNextToken())
+		{
+			m_scanner->next();
+			fatalParserError(string("Contract expected."));
+		}
+		else
+			if (Token::Contract ==  m_scanner->peekNextToken())
+				abstract = true;
+			else
+				parserError(string("Only contracts can be abstract."));
```

Comment:
Maybe move this to the ContractLevelChecker instead? Then you could also have a check for parsing the `abstract` keyword in another step before this switch and not modify the switch at all - then you would also not need any peek-calls.

Reply:
Hmm.. I'm not sure that I understood this correctly.. do you mean that the part of parsing of the `abstract` keyword should be moved to the `ContractLevelChecker`? So that the `ContractLevelChecker` is responsible for parsing the `abstract` keyword?

Reply:
I meant parse `abstract` here and do not check whether the next token is contract/interface, etc and only check that later in the contractlevelchecker

Reply:
Ah yes!